### ISSUE TITLE
hungs at training start

### ISSUE THREAD

**Issue by xv44586** [EleutherAI/gpt-neox#672]
**Describe the bug**
when I try to train a small model, first I set batch size = 32 and every GPU only uses 50% memory, so I try to upper batch size to 64, and it's hungs at the begining of training(train_step(xxx)) and not go in for hours until I kill...the GPU looks find and no more log..
but when I test it at a tiny dataset it turns all ok, ps: my whole dataset about 140G .bin
**To Reproduce**
1. use the configs/small.yml and change the batch_size = 64
2. test it at a big dataset about 140G .bin

**Expected behavior**

**Proposed solution**
I doubt the dataloader has some issue but I cannot fine where

**Screenshots**

**Environment (please complete the following information):**
 - GPUs: 8 * 40G A100
- Configs:
`# GPT-2 pretraining setup
{
   # parallelism settings ( you will want to change these based on your cluster setup, ideally scheduling pipeline stages
   # across the node boundaries )
   "pipe-parallel-size": 1,
   "model-parallel-size": 1,

   # model settings
   "num-layers": 12,
   "hidden-size": 768,
   "num-attention-heads": 12,
   "seq-length": 1024,
   "max-position-embeddings": 1024,
   "norm": "layernorm",
   "pos-emb": "rotary",
   "no-weight-tying": true,

   # these should provide some speedup but takes a while to build, set to true if desired
   "scaled-upper-triang-masked-softmax-fusion": false,
   "bias-gelu-fusion": false,

   # optimizer settings
   "optimizer": {
     "type": "Adam",
     "params": {
       "lr": 0.0006,
       "betas": [0.9, 0.95],
       "eps": 1.0e-6,
     }
   },
   "zero_optimization": {
    "stage": 0,
    "allgather_partitions": True,
    "allgather_bucket_size": 500000000,
    "overlap_comm": True,
    "reduce_scatter": True,
    "reduce_bucket_size": 500000000,
    "contiguous_gradients": True,
    "cpu_offload": False
  },

   # batch / data settings
   "train_micro_batch_size_per_gpu": 32,  # change it bigger will hung
   "gradient_accumulation_steps": 10,
   "data-impl": "mmap",
   "split": "949,50,1",

   # activation checkpointing
   "checkpoint-activations": true,
   "checkpoint-num-layers": 1,
   "partition-activations": true,
   "synchronize-each-layer": true,

   # regularization
   "gradient_clipping": 1.0,
   "weight-decay": 0.0,
   "hidden-dropout": 0.0,
   "attention-dropout": 0.0,

   # precision settings
   "fp16": {
     "enabled": true,
     "loss_scale": 0,
     "loss_scale_window": 1000,
     "hysteresis": 2,
     "min_loss_scale": 1
   },

   # misc. training settings
   "train-iters": 320000,
   "lr-decay-iters": 320000,
   "distributed-backend": "nccl",
   "lr-decay-style": "cosine",
   "warmup": 0.01,
   "save-interval": 2000,
   "eval-interval": 1000,
   "eval-iters": 10,

   # logging
   "log-interval": 100,
   "steps_per_print": 10,
   "keep-last-n-checkpoints": 4,
   "wall_clock_breakdown": true,
}
`
**Additional context**
